Add Product Attribute Code and Product Attribute Global Item Code

ahilali opened the pull request.

bguillot committed:
[ADD] Add Product Attribute Code

[ADD] Add Porduct Attribute Global Item Code

[FIX] Problems in openerp file

[ADD] README

[ADD] README

[FIX] modify openerp and product

[FIX] Sql Constraints

[FIX] Sql Constraints v2

[FIX] Product Attribute Global Item Code

[FIX] problems with openerp and init files

[FIX] problems with product file

[DEL] delete Product Variant inactive

[FIX] README image

[FIX] readme Image

[FIX] readme Image V2

[FIX] readme Image V2

[FIX] readme Image V2

[FIX] regenerate and update fr.po

[FIX] i18n and add comments

[ADD] Odoo Community Association (OCA) as author in the list and cut image

[FIX] help

[ADD] help in product file

[ADD] help in product file

[MOD] modify init file

[FIX] delete Blank line

[FIX] error code

[MOD] modify Readme file

[fix] product file

[FIX]  blank lines

[MOD] product file

[add] handle in view file

[FIX] product_attribute_global_item_code

[FIX] problem code and rename use_manual_internal_ref field

[FIX] remane use_manual_ref to use_manual_internal_ref

[FIX] merge global_code with code models

[DEL] delete product_attribute_code

[FIX] rename product builder code

[IMP] improve product file

[IMP] improve product code builder

[ADD] auto_install and installable in openerp file

[IMP] blank line at end of file

[FIX] Readme and product files

gurneyalex commented:
ECLA checked

ghost commented:
Please make 1 pull request per module

ahilali commented:
These modules ,the one depends on the other, so I made one pull request

ghost commented:
@ahilali Then make a first pull request for the first module in the dependency sequence.

bealdav commented:
@dufresnedavid I think it's not mandatory and it makes feature review and travis test complex. Not useful for linked modules, I think

dufresnedavid commented on product_attribute_global_item_code/__openerp__.py:
```diff
@@ -0,0 +1,43 @@
+
+    'name': "Product",
+
+    'summary': """Manage trainings""",
```
Manage trainings ?

dufresnedavid commented on product_attribute_global_item_code/__openerp__.py:
```diff
@@ -0,0 +1,43 @@
+
+    'summary': """Manage trainings""",
+    'description': """
+            Global Porduct Code
```
Porduct -> Product

ghost commented:
@bealdav @ahilali Ok, would you mind adding readme files with description of these modules with their purposes.

bealdav commented on product_attribute_code/__openerp__.py:
```diff
@@ -0,0 +1,38 @@
+    'version': '0.1',
+    'category': 'Generic Modules/Others',
+    'license': 'AGPL-3',
+    'description': """empty""",
```
Remove description and add template module

bealdav commented on product_attribute_code/i18n/fr.po:
```diff
@@ -0,0 +1,74 @@
+msgstr "Séquence"
+
+#~ msgid "Is Operation"
+#~ msgstr "Est une Opération"
```
Please regenerate and update this file

bealdav commented on product_attribute_code/i18n/product_attribute_code.pot:
```diff
@@ -0,0 +1,38 @@
+msgstr ""
+"Project-Id-Version: Odoo Server 8.0\n"
+"Report-Msgid-Bugs-To: \n"
+"POT-Creation-Date: 2015-02-26 15:49+0000\n"
```
File to regenerate to

bealdav commented on product_attribute_global_item_code/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+
+.. image:: https://raw.githubusercontent.com/akretion/product-attribute/8.0-variant-code/product_attribute_global_item_code/img/img_1.png
+
```
should be a relative path, thanks

bealdav commented on product_attribute_global_item_code/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+
+.. image:: https://raw.githubusercontent.com/akretion/product-attribute/8.0-variant-code/product_attribute_global_item_code/img/img_0.png
+
```
relative path

bealdav commented on product_attribute_global_item_code/__openerp__.py:
```diff
@@ -0,0 +1,40 @@
+
+# -*- coding: utf-8 -*-
+
```
why blank line here ?

bealdav commented on product_attribute_global_item_code/__openerp__.py:
```diff
@@ -0,0 +1,40 @@
+    'name': "Product",
+    'description': """
+            Global Product Code
+    """,
```
Remove description, you defined description template

bealdav commented on product_attribute_global_item_code/__openerp__.py:
```diff
@@ -0,0 +1,40 @@
+    ],
+    'demo': [
+        ],
+}
```
cut image below, to make it more readable, reduce large size to be more readable

bealdav commented on product_variant_inactive/__init__.py:
```diff
@@ -0,0 +1,11 @@
+#  Copyright (C) 2015 Akretion (http://www.akretion.com).
+#  @author Abdessamad HILALI <abdessamad.hilali@akretion.com>
+#
+##############################################################################
```
no licence here because no code => only import

bealdav commented on product_variant_inactive/__openerp__.py:
```diff
@@ -0,0 +1,40 @@
+
+# -*- coding: utf-8 -*-
+
```
blank lines

bealdav commented on product_variant_inactive/__openerp__.py:
```diff
@@ -0,0 +1,40 @@
+    'name': "Product",
+    'description': """
+            Product Invraint
+    """,
```
No description here

bealdav commented on product_variant_inactive/__openerp__.py:
```diff
@@ -0,0 +1,40 @@
+        'product_variants.xml'
+    ],
+    'demo': [
+        ],
```
please add a demo file, travis play demo data => first level test

bealdav commented on product_variant_inactive/product_variants.py:
```diff
@@ -0,0 +1,35 @@
+
```
Crop image above to be more readeable on the real use case

bealdav commented on product_variant_inactive/product_variants.py:
```diff
@@ -0,0 +1,35 @@
+class ProductVariants(models.Model):
+    """Product Variants Class"""
+    _inherit = 'product.product'
+
```
Why class name != _inherit except camel case ?
Useful doc string ?

bealdav commented on product_variant_inactive/README.rst:
```diff
@@ -0,0 +1,45 @@
+
+* https://www.odoo.com/forum/help-1
+
+
```
No block 'Installation' required => it seems the information here have no added value

bealdav commented on product_attribute_code/product.py:
```diff
@@ -0,0 +1,49 @@
+    code = fields.Char('Code')
+    comment = fields.Text('Comment')
+    readonly = fields.Boolean()
+
```
Explain with a help arg, what is the use of 'readonly' field

bealdav commented on product_attribute_global_item_code/README.rst:
```diff
@@ -0,0 +1,53 @@
+ :alt: License: AGPL-3
+
+Module name
+===========
```
Module name => Product Attribute Global Item Code

bealdav commented on product_attribute_global_item_code/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+This module was written to extend the functionality of Product
+and allow you to create a product code(default_code)
+For example :
```
... to automate default_code generation base on attributes ?

bealdav commented on product_attribute_global_item_code/__init__.py:
```diff
@@ -0,0 +1,11 @@
+#  @author Abdessamad HILALI <abdessamad.hilali@akretion.com>
+#
+##############################################################################
+
```
No licence here

bealdav commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,76 @@
+    _inherit = "product.template"
+
+    base_code = fields.Char('Code')
+
```
help arg on field, please

bealdav commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,76 @@
+    manual_default_code = fields.Char()
+    default_code = fields.Char(compute="_compute_default_code",
+                               inverse="_set_manual_default_code",
+                               store=True)
```
help arg on fields please

bealdav commented:
flake 8 fail
======== Testing test_flake8 ========

./product_attribute_code/__init__.py:24:1: W391 blank line at end of file

./product_attribute_code/__openerp__.py:38:1: W391 blank line at end of file

./product_attribute_code/product.py:49:1: E122 continuation line missing indentation or outdented

./product_attribute_global_item_code/product.py:35:80: E501 line too long (81 > 79 characters)

./product_attribute_global_item_code/product.py:40:5: E301 expected 1 blank line, found 0

ghost commented:
@ahilali consider adding unit tests, so that the coverage does not decrease

gurneyalex commented on product_attribute_global_item_code/__openerp__.py:
```diff
@@ -0,0 +1,35 @@
+
+{
+    'name': "Product",
+    'author': "Akretion",
```
please add Odoo Community Association (OCA) as author in the list.

bealdav commented on product_attribute_global_item_code/README.rst:
```diff
@@ -0,0 +1,40 @@
+
+Product Attribute Global Item Code
+===========
+
```
Missing chars = here

bealdav commented on product_attribute_code/__init__.py:
```diff
@@ -0,0 +1,24 @@
+#
+###############################################################################
+
+from . import product
```
Is text above required ?

bealdav commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,81 @@
+
+    @api.multi
+    def _build_code(self, vals):
+        self.ensure_one()
```
Put a doc string to indicate that this is method to customize: "inherit this method to customize your own behavior"

bealdav commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,81 @@
+        for value in self.attribute_value_ids:
+            if not value.attribute_id.code:
+                raise Warning(
+                    _("Missing code for %s to correct this error you must "
```
Do use Warning 
https://www.odoo.com/documentation/8.0/reference/guidelines.html#python

bealdav commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,81 @@
+
+    use_internal_ref = fields.Boolean('Manual Internal Reference',
+                                      default=False)
+    manual_default_code = fields.Char()
```
help missing

gurneyalex commented:
./product_attribute_code/**init**.py:3:1: W391 blank line at end of file

./product_attribute_code/**openerp**.py:37:1: W391 blank line at end of file

./product_attribute_code/product.py:51:1: E122 continuation line missing indentation or outdented

gdgellatly commented:
Hi clearly we need to resolve the red tests.  But a few comments.

NOTE FOR SELF: When the red has been resolved and can be built, testing the changing of the sequence of attributes changes the code or not.

I am not sure if this belongs in product-attribute or product-variant.  My feeling is that it actually belongs in product variant, and if it doesn't then I'd question whether we should keep that repository at all.  I think we will need a lot of similar modules to soemhow return the best functionality of the old product_variant_multi

It could use some tests.

gdgellatly commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,84 @@
+                      "see :Configuration  -> Product Categories -> Attribute")
+                    % (value.attribute_id.name))
+            if not value.code:
+                raise Warning(
```
This function field cause an issue with existing data.  You end up in a circle.  If you add a code to the attribute you cannot save because it wants a code on the values.  Likewise if you add on the values, it fails because you require a code on the attribute.

ahilali replied:
i totally agree with you

gdgellatly commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,84 @@
+                      "see :Configuration -> Product Categories -> "
+                      "Attribute Value exactly in %s")
+                    % (value.name, value.attribute_id.name))
+            vals.append((value.attribute_id.code, value.code))
```
I don't see why value.attribute_id.code needs to be specified anyway.  It is commonly not used in part codes.  e.g
TSHBLKXL is just as descriptive as Tsh-Co_Black-Si-XL.  So skip the raise Warning for empty value.attribute_id.code and just have

``` python
    vals.append((value.attribute_id.code or '', value.code))
```

gdgellatly commented:
In general in testing, the tight constraints of _get_default_code and _build_code combine with a lack of defaults for base_code I think needs some documenting or changing for use on existing installations.

For example, base_code must be set first on all products with that attribute.  Even then there will be a failure for existing data if more than one attribute value exists for a product.  The only way I have got this module to work with an existing data set is to comment out all the warnings, set base code first, then set attribute codes.  Then uncomment the warnings.  The average user won't do this I think.

I think also for that same reason the sql_constraints (given this is an optional field) probably need to ensure uniqueness only on NOT NULLS) and sequence might want to have a default value.

I also note that while there is a sequence field it does nothing as order is not set.

gdgellatly commented on product_attribute_code/product_view.xml:
```diff
@@ -0,0 +1,47 @@
+            <field name="inherit_id" ref="product.attribute_tree_view"/>
+            <field name="arch" type="xml">
+                <field name="name" position="before">
+                    <field name="sequence"/>
```
consider adding widget="handle" to this tag.  Unless there is a really good reason for exposing sequence numbers to users try to avoid.  Have tested and works well in editable tree view

bealdav replied:
+1

ahilali replied:
@gdgellatly first of all , thank you for your comments , please I fixed these problems you can make another review

gdgellatly replied:
@ahilali Hi,

I've taken another look but I still can't test it against a real dataset because of the circular issue.  I'm not sure how to solve it.  Maybe it needs a configuration wizard or something to set things up if there are existing attribute lines.  

Have added a couple of new comments of things I saw when reviewing the code more closely.

gdgellatly commented on product_attribute_global_item_code/product.py:
```diff
@@ -0,0 +1,86 @@
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    use_internal_ref = fields.Boolean('Manual Internal Reference',
```
consider renaming to use_manual_ref or use_manual_internal_ref . It is a bit confusing otherwise.

gdgellatly commented on product_attribute_global_item_code/product_view.xml:
```diff
@@ -0,0 +1,28 @@
+            <field name="model">product.template</field>
+            <field name="inherit_id" ref="product.product_template_only_form_view"/>
+            <field name="arch" type="xml">
+            <xpath expr="//p" position="after">
```
I think searching for all paragraph tags and relying on their only being one is a bit dangerous.  How about better qualifying it e.g. //page[@name='variants']/p or even putting using before and using the attribute_line_ids for the parse.

bealdav commented on product_code_builder/__init__.py:
```diff
@@ -0,0 +1,3 @@
+
```
empty line ?

bealdav commented on product_code_builder/__openerp__.py:
```diff
@@ -0,0 +1,39 @@
+    ],
+    'demo': [
+        ],
+    'pre_init_hook': 'update_null_and_slash_codes',
```
useful ?

bealdav commented on product_code_builder/models/product_code.py:
```diff
@@ -0,0 +1,63 @@
+from openerp import fields, models
+
+
+def update_null_and_slash_codes(cr):
```
useful ?

gdgellatly replied:
I am confused,  I wrote that for product_sequence.  It shouldn't be here.
Maybe a rebase gone wrong

On Thu, 2 Jul 2015 9:13 PM David Beal notifications@github.com wrote:

> In product_code_builder/models/product_code.py
> https://github.com/OCA/product-attribute/pull/70#discussion_r33759639:
> 
> > +#   License, or (at your option) any later version.
> > +#
> > +#   This program is distributed in the hope that it will be useful,
> > +#   but WITHOUT ANY WARRANTY; without even the implied warranty of
> > +#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> > +#   GNU Affero General Public License for more details.
> > +#
> > +#   You should have received a copy of the GNU Affero General Public License
> > +#   along with this program.  If not, see http://www.gnu.org/licenses/.
> > +#
> > +###############################################################################
> > +
> > +from openerp import fields, models
> > +
> > +
> > +def update_null_and_slash_codes(cr):
> 
> useful ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/product-attribute/pull/70/files#r33759639.

bealdav replied:
No problem, it's my fault. I asked @ahilali to copy your code here to pass runbot. But it was a bad idea.
He'll remove it. Thanks

gdgellatly replied:
Just rebase and it will pass.  Squash a few commits while you are at it ;) 

Sent from TypeMail

On 2 Jul 2015 9:49 PM, at 9:49 PM, David Beal notifications@github.com wrote:

> > +#   License, or (at your option) any later version.
> > +#
> > +#   This program is distributed in the hope that it will be useful,
> > +#   but WITHOUT ANY WARRANTY; without even the implied warranty of
> > +#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> > +#   GNU Affero General Public License for more details.
> > +#
> > +#   You should have received a copy of the GNU Affero General Public
> > License
> > +#   along with this program.  If not, see
> > http://www.gnu.org/licenses/.
> > +#
> > 
> > +###############################################################################
> > +
> > +from openerp import fields, models
> > +
> > +
> > +def update_null_and_slash_codes(cr):
> 
> No problem, it's my fault. I asked @ahilali to copy your code here to
> pass runbot. But it was a bad idea.
> He'll remove it. Thanks
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/OCA/product-attribute/pull/70/files#r33762439

ahilali commented:
@bealdav @gdgellatly @gurneyalex  thank you for review

ahilali committed:
[IMP] improve product code builder

sebastienbeau commented on product_code_builder/models/product.py:
```diff
@@ -0,0 +1,77 @@
+        'Base Code',
+        help="this field is used like a base to automatically create "
+             "Internal Reference (default_code)")
+    auto_default_code = fields.Boolean('Manual Internal Reference',
```
Could you rename the "Manual Internal Reference" by "Auto Generate Reference"?

sebastienbeau committed:
[FIX] as the field have been rename (to the opposite) the code should be updated, and also the field name

sebastienbeau commented:
@ahilali Can you update the picture, the module have change and so now there are obsolete. Thanks

sebastienbeau committed:
[FIX] simplify build code and FIX the recompute when changing the base_code, wrong value in the depends

sebastienbeau committed:
[REF] rename product_code into product_attribute

sebastienbeau commented on product_code_builder/models/product_attribute.py:
```diff
@@ -0,0 +1,48 @@
+
+    code = fields.Char('Code')
+    comment = fields.Text('Comment')
+    readonly = fields.Boolean(
```
Do not forget to remove this feature here. Should be done in an extra module

bealdav committed:
[FIX] remove readonly field

bealdav committed:
[FIX] _get_default_code

ahilali commented:
@gdgellatly  thank you for review

legalsylvain commented on product_code_builder/__openerp__.py:
```diff
@@ -0,0 +1,38 @@
+###############################################################################
+
+{
+    'name': "Product",
```
"Product" is not precise enough for a module name.

Commit:
[IMP] product Code Builder

bealdav commented:
:+1:

sebastienbeau commented on product_code_builder/models/product.py:
```diff
@@ -0,0 +1,71 @@
+        """ this method used to create a list of code elements  """
+        self.ensure_one()
+        res = self.base_code or ''
+        for value in self.attribute_value_ids:
```
After looking twice at that code, I pretty sure that we have an issue here.
Indeed if I have 2 attributes Color and Size. I have set the Color with the sequence 1 and the Size with the sequence 2. The expected behaviour is to build a code always with the color at the start. But as we do not have any sort and just take the self.attribute_value_ids, I think that we can have case where the order is wrong. What do you think?

rvalyi replied:
@sebastienbeau @renatonlima and me had the same doubt here. In the old OCA module we made in v7, we had some sorting I believe. product.attribute.value have a sequence, but product.attribute don't have one so you added one in this module (good point). Probably having a _order = "sequence" in the ProductAttribute class in this module would do the trick (to be verified). Just like https://github.com/odoo/odoo/blob/8.0/addons/product/product.py#L357 but for the sequence field you added for ProductAttribute then.

ahilali replied:
@rvalyi  thanks

Commit:
[FIX] Product Code

bealdav committed:
[IMP] build product code according to the order of the attribute code

bealdav commented:
WORK IN PROGRESS

bealdav committed:
[IMP] add test unit, simplify code, documentation

Commit:
Improve module product_code_builder

Fix tests
Add documentation
Translate and Pep8 in test_code.py
Add incompatibilty with product_sequence module
Change behavior of auto

Auto = On -> generate default_code
Auto = Off -> use only user provided default_code

Previously prefix was used with auto == off

hparfr committed:
Update documentation

sebastienbeau commented:
@david runbot is green because of installing other module (data or demo) what do you think about setting False as default value. So we will have no issue when installing the module with existing data, and overwriting the default method to return True if we are not in install mode?

sebastienbeau committed:
[FIX] fix installation, set False for existing record

sebastienbeau committed:
[FIX] when installing the module the existing code must be kept

sebastienbeau committed:
[FIX] fix test

sebastienbeau committed:
[FIX] fix dependency, only need to depend of product

sebastienbeau committed:
[IMP] add uniq constraint on prefix code

sebastienbeau committed:
[FIX] fix edition of default code for none varianted product and also fix compatibility with other module

sebastienbeau committed:
[IMP] initialize the prefix code by the default code of the product

sebastienbeau committed:
[ADD] add clue module for product sequence and product code builder

sebastienbeau committed:
[FIX] modifiy demo data to fix issue when testing

sebastienbeau committed:
[FIX] fix pep 8

sebastienbeau commented:
Wouhou test are green

sebastienbeau committed:
[IMP] improve UI and help

bealdav commented:
nice, ready for merge, used by 2 of our clients in production

mourad-ehm commented:
:+1:  code review

sebastienbeau committed:
[REF] simplify code to compute the default code, use order on attribute value for computing the default code

sebastienbeau commented:
I have done some refactor to simplify the code and add some extra test

florian-dacosta commented:
:+1: code review
